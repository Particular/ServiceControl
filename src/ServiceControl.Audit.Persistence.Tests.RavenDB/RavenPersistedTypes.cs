namespace ServiceControl.UnitTests
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using NServiceBus;
    using NUnit.Framework;
    using Particular.Approvals;
    using Raven.Abstractions.Data;
    using Raven.Client.Indexes;
    using ServiceControl.Audit.Auditing;
    using ServiceControl.Audit.Monitoring;
    using ServiceControl.Audit.Persistence.RavenDb;
    using ServiceControl.Audit.Persistence.Tests;

    class RavenPersistedTypes : PersistenceTestFixture
    {
        [Test]
        public void Verify()
        {
            var ravenPersistenceType = typeof(RavenDbPersistenceConfiguration);

            var ravenPersistenceTypes = ravenPersistenceType.Assembly.GetTypes()
                .Where(type => typeof(AbstractIndexCreationTask).IsAssignableFrom(type))
                .SelectMany(indexType => indexType.BaseType?.GenericTypeArguments)
                .Distinct();

            var documentTypeNames = string.Join(Environment.NewLine, ravenPersistenceTypes.Select(t => t.AssemblyQualifiedName));

            Approver.Verify(documentTypeNames);
        }

        [Test]
        public async Task CanLoadLegacyTypes()
        {
            var messageId = Guid.NewGuid().ToString();
            var endpointName = "Sales";
            var message = MakeMessage(messageId: messageId, processingEndpoint: endpointName);
            message.MessageMetadata.Add("ReceivingEndpoint", new EndpointDetails()
            {
                Host = "sender@machine",
                HostId = Guid.NewGuid(),
                Name = "Sender"
            });

            await IngestProcessedMessagesAudits(message);

            _ = await configuration.DocumentStore.AsyncDatabaseCommands.PatchAsync(message.Id, new ScriptedPatchRequest()
            {
                Script =
@"
var messageMetadata = this.MessageMetadata;
messageMetadata['ReceivingEndpoint']['$type'] = 'This.Is.The.Wrong.EndpointDetails, InTheWrongAssembly';
"
            });

            await configuration.CompleteDBOperation();

            var messages = await DataStore.GetMessages(false, new Audit.Infrastructure.PagingInfo(), new Audit.Infrastructure.SortInfo("Id", "asc"));

            Assert.That(messages, Is.Not.Null);
            Assert.That(messages.Results, Is.Not.Null);
            Assert.That(messages.Results.Single().Id, Is.EqualTo(message.UniqueMessageId));
        }

        ProcessedMessage MakeMessage(
            string messageId = null,
            MessageIntentEnum intent = MessageIntentEnum.Send,
            string conversationId = null,
            string processingEndpoint = null
        )
        {
            messageId ??= Guid.NewGuid().ToString();
            conversationId ??= Guid.NewGuid().ToString();
            processingEndpoint ??= "SomeEndpoint";

            var metadata = new Dictionary<string, object>
            {
                { "MessageId", messageId },
                { "MessageIntent", intent },
                { "CriticalTime", TimeSpan.FromSeconds(5) },
                { "ProcessingTime", TimeSpan.FromSeconds(1) },
                { "DeliveryTime", TimeSpan.FromSeconds(4) },
                { "IsSystemMessage", false },
                { "MessageType", "MyMessageType" },
                { "IsRetried", false },
                { "ConversationId", conversationId },
                //{ "ContentLength", 10}
            };

            var headers = new Dictionary<string, string>
            {
                { Headers.MessageId, messageId },
                { Headers.ProcessingEndpoint, processingEndpoint },
                { Headers.MessageIntent, intent.ToString() },
                { Headers.ConversationId, conversationId }
            };


            return new ProcessedMessage(headers, metadata);
        }

        async Task IngestProcessedMessagesAudits(params ProcessedMessage[] processedMessages)
        {
            var unitOfWork = StartAuditUnitOfWork(processedMessages.Length);
            foreach (var processedMessage in processedMessages)
            {
                await unitOfWork.RecordProcessedMessage(processedMessage);
            }
            await unitOfWork.DisposeAsync();
            await configuration.CompleteDBOperation();
        }
    }
}