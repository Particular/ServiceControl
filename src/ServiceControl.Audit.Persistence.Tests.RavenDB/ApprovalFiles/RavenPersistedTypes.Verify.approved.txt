ServiceControl.SagaAudit.SagaHistory, ServiceControl.Audit.Persistence.SagaAudit, Version=4.0.0.0, Culture=neutral, PublicKeyToken=null
ServiceControl.Audit.Persistence.Monitoring.KnownEndpoint, ServiceControl.Audit.Persistence, Version=4.0.0.0, Culture=neutral, PublicKeyToken=null
ServiceControl.Audit.Auditing.ProcessedMessage, ServiceControl.Audit.Persistence, Version=4.0.0.0, Culture=neutral, PublicKeyToken=null
ServiceControl.Audit.Auditing.FailedAuditImport, ServiceControl.Audit.Persistence, Version=4.0.0.0, Culture=neutral, PublicKeyToken=null
ServiceControl.Audit.Persistence.RavenDb.Indexes.MessagesViewIndex+SortAndFilterOptions, ServiceControl.Audit.Persistence.RavenDb, Version=4.0.0.0, Culture=neutral, PublicKeyToken=null